### PR TITLE
[Modules] Implemented CMK for Storage Account

### PR DESCRIPTION
**Pull request by MrMCake**
# Description

- Implemented CMK for Storage Account
- Note: This will only work with a UserAssignedIdentity. To also enable the storage account's system assigned identity, the storage account deployment would need to be duplicated (i.e. create storage account without encryption but with MSI, then create it again with encryption using the MSI that with granted access to the key vault. Maybe something to think about for later)
- For some reason it is not possible to deploy encryption with both a user assigned identity & system assigned identity being part of the deployment. As soon as you do this, the encryption throws an error that the user assigned identity is invalid. I noted this in the `cMKKeyName` parameter
- Added an additional parameter file to test CMK

| Pipeline |
| - |
|[Storage: StorageAccounts](https://github.com/Azure/ResourceModules/actions/workflows/ms.storage.storageaccounts.yml) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

### CODE REVIEW

File: arm/.global/global.module.tests.ps1 [hunk 1]
```diff
@@ -451,7 +451,12 @@ Describe 'Readme tests' -Tag Readme {
             $fileHashAfter = (Get-FileHash $readMeFilePath).Hash
 
             # Compare
-            $fileHashBefore -eq $fileHashAfter | Should -Be $true -Because 'The file hashes before and after applying the Set-ModuleReadMe function should be identical'
+            $filesAreTheSame = $fileHashBefore -eq $fileHashAfter
```
Comment (eriqua) [inline arm/.global/global.module.tests.ps1:454]: Is this related or should we add it to the global testing PR?

File: arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json [hunk 2]
```diff
@@ -5,22 +5,21 @@
         "name": {
             "value": "<<namePrefix>>-wd-aut-encr-001"
         },
-        "encryptionKeySource" : {
-            "value" : "Microsoft.Keyvault"
+        "encryptionKeySource": {
```
Comment (eriqua) [inline arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json:8]: There's no reference of the automation account changes in the PR title and description. Are these changes intended to be included here?

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep [hunk 3]
```diff
@@ -189,16 +206,30 @@ resource storageAccount 'Microsoft.Storage/storageAccounts@2021-08-01' = {
...
+        }
       }
       requireInfrastructureEncryption: storageAccountKind != 'Storage' ? requireInfrastructureEncryption : null
+      keyvaultproperties: !empty(cMKeyName) ? {
+        keyname: cMKeyName
+        keyvaulturi: keyVault.properties.vaultUri
+        keyversion: !empty(cMKeyVersion) ? cMKeyVersion : null
+      } : null
+      identity: !empty(cMKeyName) ? {
+        userAssignedIdentity: cMKUserAssignedIdenityResourceId
```
Comment (eriqua) [inline arm/Microsoft.Storage/storageAccounts/deploy.bicep:231]:
```suggestion
        userAssignedIdentity: cMKUserAssignedIdentityResourceId
```

File: arm/Microsoft.Storage/storageAccounts/.parameters/encr.parameters.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+                "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/adp-<<namePrefix>>-az-msi-x-001": {}
+            }
+        },
+        "cMKKeyVaultResourceId": {
+            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.KeyVault/vaults/adp-<<namePrefix>>-az-kv-nopr-002"
+        },
+        "cMKeyName": {
+            "value": "keyEncryptionKey"
+        },
+        "cMKUserAssignedIdenityResourceId": {
```
Comment (eriqua) [inline arm/Microsoft.Storage/storageAccounts/.parameters/encr.parameters.json:34]:
```suggestion
        "cMKUserAssignedIdentityResourceId": {
```

File: arm/.global/global.module.tests.ps1 (hunk 1 above)
Comment (MrMCake) [reply arm/.global/global.module.tests.ps1:454]: I have this now already in 3 different PRs as I need it to be able to troubleshoot the ReadMe generation. So yes - it is also in that one. ;) It's just an adjustment to the output as before it was really hard to say which readme generation failed if you have it running for all 85

File: arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json (hunk 2 above)
Comment (MrMCake) [reply arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json:8]: The fix for the indent? Should I create a new PR for that? It was an automatic fix when I saved the file.

Comment (MrMCake) [reply arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json:8]: I think just the incorrect 'carml' in the keyvaulturi was an actually fix

File: arm/Microsoft.Storage/storageAccounts/.parameters/encr.parameters.json (hunk 4 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/.parameters/encr.parameters.json]: fixed

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep (hunk 3 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/deploy.bicep]: fixed

File: arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json (hunk 2 above)
Comment (MrMCake) [reply arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json:8]: By the way, the reason I even touched the automation account in the first place was that I used its CMK implementation as inspiration ;)

File: arm/.global/global.module.tests.ps1 (hunk 1 above)
Comment (eriqua) [reply arm/.global/global.module.tests.ps1:454]: I'm not discussing the value of it, but please, especially if included in another PR, remove this change from here.

File: arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json (hunk 2 above)
Comment (eriqua) [reply arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json:8]: Again I'm not discussing the value of it but generally kindly asking to make the review process easier. Nothing is mentioned in the PR description about this change so I'm not expecting the automation account to have updates. Could you please add one line in the PR description?

File: arm/.global/global.module.tests.ps1 (hunk 1 above)
Comment (MrMCake) [reply arm/.global/global.module.tests.ps1:454]: It's not included. It was just something to make this PR work - and others after. Will create another PR for it

Comment (MrMCake) [reply arm/.global/global.module.tests.ps1]: Removed

File: arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json (hunk 2 above)
Comment (MrMCake) [reply arm/Microsoft.Automation/automationAccounts/.parameters/encr.parameters.json]: Removed

File: arm/Microsoft.Storage/storageAccounts/.parameters/encr.parameters.json [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        },
+        "userAssignedIdentities": {
+            "value": {
+                "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/adp-<<namePrefix>>-az-msi-x-001": {}
+            }
+        },
+        "cMKKeyVaultResourceId": {
+            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.KeyVault/vaults/adp-<<namePrefix>>-az-kv-nopr-002"
+        },
+        "cMKeyName": {
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/.parameters/encr.parameters.json:31]: Just looking at the other parameters, should this be "prefixed" `cMK` as well?
```suggestion
        "cMKKeyName": {
```

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep [hunk 6]
```diff
@@ -138,6 +138,18 @@ param diagnosticMetricsToEnable array = [
   'Transaction'
 ]
 
+@description('Optional. The resource ID of a key vault to reference a customer managed key for encryption from.')
+param cMKKeyVaultResourceId string = ''
+
+@description('Optional. The name of the customer managed key to use for encryption. Cannot be deployed together with the parameter \'systemAssignedIdentity\' enabled.')
+param cMKeyName string = ''
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/deploy.bicep:145]: Same comment as before.
```suggestion
param cMKKeyName string = ''
```

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep [hunk 7]
```diff
@@ -138,6 +138,18 @@ param diagnosticMetricsToEnable array = [
...
+
+@description('Conditional. User assigned identity to use when fetching the customer managed key. Required if \'cMKeyName\' is not empty.')
+param cMKUserAssignedIdentityResourceId string = ''
+
+@description('Optional. The version of the customer managed key to reference for encryption. If not provided, latest is used.')
+param cMKeyVersion string = ''
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/deploy.bicep:151]:
```suggestion
param cMKKeyVersion string = ''
```

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep [hunk 8]
```diff
@@ -189,16 +206,30 @@ resource storageAccount 'Microsoft.Storage/storageAccounts@2021-08-01' = {
   tags: tags
   properties: {
     encryption: {
-      keySource: 'Microsoft.Storage'
+      keySource: !empty(cMKeyName) ? 'Microsoft.Keyvault' : 'Microsoft.Storage'
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/deploy.bicep:209]:
```suggestion
      keySource: !empty(cMKKeyName) ? 'Microsoft.Keyvault' : 'Microsoft.Storage'
```

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep [hunk 9]
```diff
@@ -189,16 +206,30 @@ resource storageAccount 'Microsoft.Storage/storageAccounts@2021-08-01' = {
...
         } : null
+        table: {
+          enabled: true
+        }
+        queue: {
+          enabled: true
+        }
       }
       requireInfrastructureEncryption: storageAccountKind != 'Storage' ? requireInfrastructureEncryption : null
+      keyvaultproperties: !empty(cMKeyName) ? {
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/deploy.bicep:225]:
```suggestion
      keyvaultproperties: !empty(cMKKeyName) ? {
```

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep [hunk 10]
```diff
@@ -189,16 +206,30 @@ resource storageAccount 'Microsoft.Storage/storageAccounts@2021-08-01' = {
...
+        keyname: cMKeyName
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/deploy.bicep:226]:
```suggestion
        keyname: cMKKeyName
```

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep [hunk 11]
```diff
@@ -189,16 +206,30 @@ resource storageAccount 'Microsoft.Storage/storageAccounts@2021-08-01' = {
...
+        keyvaulturi: keyVault.properties.vaultUri
+        keyversion: !empty(cMKeyVersion) ? cMKeyVersion : null
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/deploy.bicep:228]:
```suggestion
        keyversion: !empty(cMKKeyVersion) ? cMKKeyVersion : null
```

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep [hunk 12]
```diff
@@ -189,16 +206,30 @@ resource storageAccount 'Microsoft.Storage/storageAccounts@2021-08-01' = {
...
+      } : null
+      identity: !empty(cMKeyName) ? {
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/deploy.bicep:230]:
```suggestion
      identity: !empty(cMKKeyName) ? {
```

File: arm/Microsoft.Storage/storageAccounts/readme.md [hunk 13]
```diff
@@ -33,12 +33,20 @@ This module is used to deploy a storage account, with the ability to deploy 1 or
...
+| :-- | :-- | :-- | :-- |
+| `cMKUserAssignedIdentityResourceId` | string | `''` | User assigned identity to use when fetching the customer managed key. Required if 'cMKeyName' is not empty. |
+
 **Optional parameters**
 | Parameter Name | Type | Default Value | Allowed Values | Description |
 | :-- | :-- | :-- | :-- | :-- |
 | `allowBlobPublicAccess` | bool | `False` |  | Indicates whether public access is enabled for all blobs or containers in the storage account. For security reasons, it is recommended to set it to false. |
 | `azureFilesIdentityBasedAuthentication` | object | `{object}` |  | Provides the identity based authentication settings for Azure Files. |
 | `blobServices` | _[blobServices](blobServices/readme.md)_ object | `{object}` |  | Blob service and containers to deploy. |
+| `cMKeyName` | string | `''` |  | The name of the customer managed key to use for encryption. Cannot be deployed together with the parameter 'systemAssignedIdentity' enabled. |
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/readme.md:47]:
```suggestion
| `cMKKeyName` | string | `''` |  | The name of the customer managed key to use for encryption. Cannot be deployed together with the parameter 'systemAssignedIdentity' enabled. |
```

File: arm/Microsoft.Storage/storageAccounts/readme.md [hunk 14]
```diff
@@ -33,12 +33,20 @@ This module is used to deploy a storage account, with the ability to deploy 1 or
...
+| `cMKeyVersion` | string | `''` |  | The version of the customer managed key to reference for encryption. If not provided, latest is used. |
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/readme.md:48]:
```suggestion
| `cMKKeyVersion` | string | `''` |  | The version of the customer managed key to reference for encryption. If not provided, latest is used. |
```

File: arm/Microsoft.Storage/storageAccounts/readme.md [hunk 15]
```diff
@@ -367,8 +375,37 @@ The hierarchical namespace of the storage account (see parameter `enableHierarch
...
+        },
+        "userAssignedIdentities": {
+            "value": {
+                "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/adp-<<namePrefix>>-az-msi-x-001": {}
+            }
+        },
+        "cMKKeyVaultResourceId": {
+            "value": "/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.KeyVault/vaults/adp-<<namePrefix>>-az-kv-nopr-002"
+        },
+        "cMKeyName": {
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/readme.md:404]:
```suggestion
        "cMKKeyName": {
```

File: arm/Microsoft.Storage/storageAccounts/readme.md [hunk 16]
```diff
@@ -385,7 +422,18 @@ The hierarchical namespace of the storage account (see parameter `enableHierarch
...
+    storageAccountSku: 'Standard_LRS'
     allowBlobPublicAccess: false
+    publicNetworkAccess: 'Disabled'
+    requireInfrastructureEncryption: true
+    systemAssignedIdentity: false
+    userAssignedIdentities: {
+      '/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/adp-<<namePrefix>>-az-msi-x-001': {}
+    }
+    cMKKeyVaultResourceId: '/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.KeyVault/vaults/adp-<<namePrefix>>-az-kv-nopr-002'
+    cMKeyName: 'keyEncryptionKey'
```
Comment (MariusStorhaug) [inline arm/Microsoft.Storage/storageAccounts/readme.md:435]:
```suggestion
    cMKKeyName: 'keyEncryptionKey'
```

**Review by MariusStorhaug — CHANGES_REQUESTED** [review]

File: arm/Microsoft.Storage/storageAccounts/.parameters/encr.parameters.json (hunk 5 above)
Comment (MariusStorhaug) [reply arm/Microsoft.Storage/storageAccounts/.parameters/encr.parameters.json:31]: `CMK` = The mode of managing encryption keys.
`Key` is an "object"/"class" of the mode
`Name` is a property of the key.
= `CMKKeyMode`

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep (hunk 6 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/deploy.bicep]: done

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep (hunk 8 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/deploy.bicep]: done

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep (hunk 12 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/deploy.bicep]: done

File: arm/Microsoft.Storage/storageAccounts/readme.md (hunk 13 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/readme.md]: done

File: arm/Microsoft.Storage/storageAccounts/readme.md (hunk 15 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/readme.md]: done

File: arm/Microsoft.Storage/storageAccounts/readme.md (hunk 16 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/readme.md]: done

File: arm/Microsoft.Storage/storageAccounts/readme.md (hunk 14 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/readme.md]: done

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep (hunk 9 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/deploy.bicep]: done

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep (hunk 7 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/deploy.bicep]: done

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep (hunk 10 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/deploy.bicep]: done

File: arm/Microsoft.Storage/storageAccounts/deploy.bicep (hunk 11 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/deploy.bicep]: done

File: arm/Microsoft.Storage/storageAccounts/.parameters/encr.parameters.json (hunk 5 above)
Comment (MrMCake) [reply arm/Microsoft.Storage/storageAccounts/.parameters/encr.parameters.json]: done

**Review by eriqua — APPROVED** [review]
🆗
